sw/device/lib/base/memory_hardened.h

```diff
@@ -0,0 +1,185 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#ifndef OPENTITAN_SW_DEVICE_LIB_BASE_MEMORY_HARDENED_H_
+#define OPENTITAN_SW_DEVICE_LIB_BASE_MEMORY_HARDENED_H_
+
+/**
+ * @file
+ * @brief OpenTitan Hardened Memory Function Library
+ *
+ * This library provides implementation of memory functions for copy, set
+ * and mask memory regions which verifies completition of the operations.
+ *
+ * Fast shuffling of memory accesses is done by choosing:
+ *  1. random stride for memory accesses
+ *  2. random starting position
+ *
+ * Stride is selected to be co-prime with length, so we can simply add it
+ * at each iteration and be sure we will go through all elements.
+ * This also gives opportunity to add decoy memory accesses.
+ *
+ * Such implementation avoids computing random permutations and still
+ * introduce enough noise, while being just 2x slower than regular memcpy().
+ */
+
+#include <stdalign.h>
+#include <stddef.h>
+#include <stdint.h>
+
+#include "sw/device/lib/base/hardened.h"
+
+#ifdef __cplusplus
+extern "C" {
+// C++ don't have `restrict` keyword
+#define restrict
+#endif  // __cplusplus
+
+/**
+ * Fault resistant memcpy(). Checks that copy was fully done.
+ *
+ * @param dest Destination buffer
+ * @param src Source buffer
+ * @param len Bytes to copy
+ * @return `kHardenedBoolTrue` if everything is ok, undefined value else.
+ *         The only guaranteed result is kHardenedBoolTrue.
+ */
+hardened_bool_t memcpy_checked(void *restrict dest, const void *restrict src,
```

Comment:
The returned bool seems important to check. Perhaps mark it with `warn_unused_result`? (See the DIFs for examples of how to achieve this.)